### PR TITLE
Ceara/aitt 329 restore rubric tester

### PR DESCRIPTION
**Pull request by cearachew**
Restores rubric testing and adds command line options for selecting which lesson(s) to run and whether to re-download lesson files. Also adds confusion matrices to the html output. Command line options are:
```
options:
  -h, --help            show this help message and exit
  --lesson-names LESSON_NAMES
                        Comma-separated list of lesson names to run. Supported lessons U3-2022-L10, U3-2022-L13, U3-2022-L17,
                        U3-2022-L20, U3-2022-L23, New-U3-2022-L10, New-U3-2022-L13. Defaults to all lessons.
  -o OUTPUT_FILENAME, --output-filename OUTPUT_FILENAME
                        Output filename within output directory
  -c, --use-cached      Use cached responses from the API.
  -l LLM_MODEL, --llm-model LLM_MODEL
                        Which LLM model to use. Supported models: gpt-4, gpt-4-0314, gpt-4-32k, gpt-4-32k-0314. Default: gpt-4
  -n NUM_RESPONSES, --num-responses NUM_RESPONSES
                        Number of responses to generate for each student. Defaults to 1.
  -p NUM_PASSING_GRADES, --num-passing-grades NUM_PASSING_GRADES
                        Number of grades which are considered passing.
  -s MAX_NUM_STUDENTS, --max-num-students MAX_NUM_STUDENTS
                        Maximum number of students to grade. Defaults to 100 students.
  --student-ids STUDENT_IDS
                        Comma-separated list of student ids to grade. Defaults to all students.
  -t TEMPERATURE, --temperature TEMPERATURE
                        Temperature of the LLM. Defaults to 0.0.
  -d, --download        re-download lesson files, overwriting previous files
```

For each lesson, it will look in `lesson_data/` for a directory with those files, and download them if that directory doesn't exist or the download option is selected. It will run the AI assessment for that lesson, and cache the results in the lesson directory (to avoid name conflicts with multiple lessons having `student_n` files), then generate an html report for the lesson.

Confusion matrices added: 
Screenshot 2023-12-06 at 6 45 12 PM

Running this script requires the `OPEN_AI_KEY` and `PYTHONPATH` environment variables to be set

### CODE REVIEW

File: lib/assessment/rubric_tester.py
Comment (cearachew) [reply lib/assessment/rubric_tester.py]: Just changed this to have the default value of write_cached be False and set this to True here

**Comment by cat5inthecradle** [comment]
Is "Predicted" and "Expected" correct? Those sound very similar. Should it be "Predicted" and "Actual"?

**Comment by cearachew** [comment]
> Is "Predicted" and "Expected" correct? Those sound very similar. Should it be "Predicted" and "Actual"?

The naming conventions in the code and html output are pretty non-standard for ML already. They currently use "Actual" for the labels returned by the LLM and "Expected" for the gold labels, which IMO is confusing. I would prefer "Predicted" for LMM labels and "Actual" or "True" for the gold labels, but that's a pretty big refactor.

File: .dockerignore
Comment (davidsbailey) [inline .dockerignore:5]: based on the 🚫 symbol github is showing us, please check your editor to make sure that you always have newlines at the end of the file, and also update this PR to includes those newlines. (apologies, I realize one or more of the files was missing one previously)

File: lib/assessment/grade.py
Comment (davidsbailey) [reply lib/assessment/grade.py:134]: I'm having trouble locating the discussion, but as we discussed offline, the service will never be configured to write cache files to the filesystem. this technique will only be used by the command line tool which is being restored in this PR.

**Comment by davidsbailey** [comment]
> > Is "Predicted" and "Expected" correct? Those sound very similar. Should it be "Predicted" and "Actual"?
> 
> The naming conventions in the code and html output are pretty non-standard for ML already. They currently use "Actual" for the labels returned by the LLM and "Expected" for the gold labels, which IMO is confusing. I would prefer "Predicted" for LMM labels and "Actual" or "True" for the gold labels, but that's a pretty big refactor.

This is my understanding of the terminology problem:
* CS testing terminology (which this codebase largely uses today)
  * expected: the "correct" value for the subject-under-test to produce on a certain input
  * actual: the value the subject-under-test does produce on a certain input
* ML terminology
  * actual: the "correct" or "true" classification for a particular input
  * predicted: the classification assigned to a particular input by an AI model

thus, CS testing "expected" is like ML "actual", and CS testing "actual" is like ML "predicted".

I know we talked about "Predicted" and "Expected", but I'm worried that using our own hybrid terminology is a bit too much of a crutch and thinking we should lean in to ML terms.

for this PR, my thoughts are:
* fix only what is visible in the report for now
* prefer ML terms over CS terms in the report
* "use more words" to help folks not accustomed to ML terms to understand them, e.g. 
  * "Predicted (by AI)" or "Predicted (AI score)", and
  * "Actual (human score)"

interested how this approach sounds to other folks, and def open to other wordings

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [inline lib/assessment/rubric_tester.py:123]: Looks like this has been merged now. Maybe worth pulling in now to make sure there are no surprises in merge conflicts later?

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [inline lib/assessment/rubric_tester.py:71]: nit: "supported models" --> "supported lessons"

File: requirements.txt
Comment (davidsbailey) [inline requirements.txt:9]: I am curious, what does scikit do?

File: requirements.txt
Comment (davidsbailey) [inline requirements.txt:10]: please pin a specific version, so that library changes do not break us. I know gdown is less likely to have breaking changes than openai, but it seems this is the pattern we have chosen for now. or if you know how to pin a major version number or another best practice that works too!

**Review by davidsbailey — APPROVED** [review]
just a few nits. thanks for getting this working again! LGTM after considering comments

File: requirements.txt
Comment (davidsbailey) [reply requirements.txt:9]: please disregard, now I see `sklearn` in the imports list

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [inline lib/assessment/rubric_tester.py:1]: could you please add some instructions to README.md to cover the basics of running this? I'm running into problems:

```
Dave-MBP:~/src/aiproxy (ceara/AITT-329-restore-rubric-tester)$ python3 -mvenv .venv
Dave-MBP:~/src/aiproxy (ceara/AITT-329-restore-rubric-tester)$ source .venv/bin/activate
(.venv) Dave-MBP:~/src/aiproxy (ceara/AITT-329-restore-rubric-tester)$ python3 -m pip install -U pip
(.venv) Dave-MBP:~/src/aiproxy (ceara/AITT-329-restore-rubric-tester)$ pip install -r requirements.txt
(.venv) Dave-MBP:~/src/aiproxy (ceara/AITT-329-restore-rubric-tester)$ python3 ./lib/assessment/rubric_tester.py --help
Traceback (most recent call last):
  File "/Users/dsb/src/aiproxy/./lib/assessment/rubric_tester.py", line 18, in <module>
    from lib.assessment.config import SUPPORTED_MODELS, VALID_GRADES, LESSONS
ModuleNotFoundError: No module named 'lib'
```

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [inline lib/assessment/rubric_tester.py:247]: one more thought: could this file be moved to the bin directory? I could be missing something, but it seems like it is more of a binary than a library

File: requirements.txt
Comment (cearachew) [reply requirements.txt:9]: I'm using sklearn's metric library to do the accuracy calculation/ confusion matrices

File: requirements.txt
Comment (cearachew) [reply requirements.txt:10]: Yeah, we can pin major (and minor) versions with `library_name>=major(.minor)`. This is usually better than absolute versions so that pip can pull patch updates and resolve library dependency issues

File: lib/assessment/rubric_tester.py
Comment (cearachew) [reply lib/assessment/rubric_tester.py:123]: I already pulled Mark's changes in, so I'll just delete this comment

**Comment by cearachew** [comment]
> > > Is "Predicted" and "Expected" correct? Those sound very similar. Should it be "Predicted" and "Actual"?
> > 
> > 
> > The naming conventions in the code and html output are pretty non-standard for ML already. They currently use "Actual" for the labels returned by the LLM and "Expected" for the gold labels, which IMO is confusing. I would prefer "Predicted" for LMM labels and "Actual" or "True" for the gold labels, but that's a pretty big refactor.
> 
> for this PR, my thoughts are:
> 
> * fix only what is visible in the report for now
> * prefer ML terms over CS terms in the report
> * "use more words" to help folks not accustomed to ML terms to understand them, e.g.
>   
>   * "Predicted (by AI)" or "Predicted (AI score)", and
>   * "Actual (human score)"
> 
> interested how this approach sounds to other folks, and def open to other wordings

Here's the report output with these changes:
Screenshot 2023-12-12 at 12 04 56 PM
Screenshot 2023-12-12 at 12 05 06 PM

File: requirements.txt
Comment (cat5inthecradle) [reply requirements.txt]: I'll add a line item to the Engineering Excellence "Manage Outdated Dependencies" project to get dependabot setup for this repo so we can more easily maintain dependency versions going forward.

File: lib/assessment/rubric_tester.py
Comment (davidsbailey) [reply lib/assessment/rubric_tester.py:1]: The new instructions are working for me now, thanks!

**Review by davidsbailey — APPROVED** [review]
Updates look good!

**Pull request merged**
